ads/inabox/inabox-messaging-host.js

```diff
@@ -209,35 +209,97 @@ export class InaboxMessagingHost {
+   *   If we have a frame hierarchy:
+   *     Host -> Friendly Frame -> X Domain Frame 1 -> Message Source Frame
+   *     and "Friendly Frame" is registered in this.iframes_, then
+   *     "Message Source Frame" is valid, because one of its parent frames
+   *     is registered in this.iframes_, and the result of the call to
+   *     getFrameElement_ would be the iframe "X Domain Frame 1" as it is
+   *     the deepest frame that the host doc can accurately measure.
+   * Note: The sentinel should be unique to the source window, and the result
    * is cached using the sentinel as the key.
    *
    * @param source {!Window}
    * @param sentinel {string}
-   * @returns {?HTMLIFrameElement}
+   * @return {?Element}
    * @private
    */
   getFrameElement_(source, sentinel) {
     if (this.iframeMap_[sentinel]) {
       return this.iframeMap_[sentinel];
     }
-
-    // Walk up on the window tree until find host's direct child window
-    while (source.parent !== this.win_ && source !== this.win_.top) {
-      source = source.parent;
-    }
-
-    // Find the iframe element that hosts the message source window
+    const measureableFrame = this.getMeasureableFrame(source);
+    const measureableWin = measureableFrame.contentWindow;
     for (let i = 0; i < this.iframes_.length; i++) {
       const iframe = this.iframes_[i];
-      if (iframe.contentWindow === source) {
-        // Cache the found iframe with its sentinel.
-        this.iframeMap_[sentinel] = iframe;
-        return iframe;
+      for (let j = 0, tempWin = measureableWin;
+        j < 10; j++, tempWin = tempWin.parent) {
+        if (iframe.contentWindow == tempWin) {
+          this.iframeMap_[sentinel] = measureableFrame;
+          return measureableFrame;
+        }
       }
     }
     return null;
   }
+
+  /**
+   * Returns whichever window in win's parent hierarchy is the deepest window
+   * that is measurable from the perspective of the current window.
+   * For when win is nested within a x-domain frame, walks up the window's
+   * parent hierarchy until the top-most x-domain frame in the hierarchy
+   * is found. Then, it returns the frame element for that window.
+   * For when win is friendly framed, returns the frame element for win.
+   * @param {!Window} win
+   * @return {?Element}
+   * @visibleForTesting
+   */
+  getMeasureableFrame(win) {
+    let measureableWin;
+    for (let j = 0, tempWin = win; j < 10 && tempWin != tempWin.top;
+      j++, tempWin = tempWin.parent) {
+      if (canInspectWindow(tempWin)) {
+        break;
+      }
+      measureableWin = tempWin;
+    }
+    if (!!measureableWin) {
+      const parent = measureableWin.parent;
+      const iframes = parent.document.querySelectorAll('iframe');
+      for (let k = 0, frame = iframes[k]; k < iframes.length;
+        k++, frame = iframes[k]) {
+        if (frame.contentWindow == measureableWin) {
+          return frame;
+        }
+      }
+    }
+    return win.frameElement;
```

Comment:
i'm very confused. did you mean `measuarableWin.frameElement`?

in fact, can we just do:
```js
function getMeasureableFrame(win) {
  if (win.parent === win) {
    return null;
  }
  if (win.frameElement) {
    return win.frameElement;
  }
  return getMeasureableFrame(win.parent);
}
```

Reply:
I should add more comments explaining why this is the way it is as I agree it's pretty confusing

The reason that this is so complicated, is that if you have a window reference to a xdomain frame, you can not do win.frameElement, even if you actually do have access to that window's frame element. So, instead we have to find the top-most x-domain window, then go up to its parent, and query select on that parent's document for all its child iframes, and then loop over them to find the correct frame element by comparing the frame element's contentWindow to the xdomain window we care about. 

However, if there is no xdomain iframe, then we can just return the source's frameElement